hare/eligibility/oracle.go

```diff
@@ -0,0 +1,106 @@
+package eligibility
+
+import (
+	"bytes"
+	"errors"
+	"github.com/davecgh/go-xdr/xdr2"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/types"
+	"hash/fnv"
+	"math"
+)
+
+const genesisLayer = types.LayerID(0) // TODO: read from conf
+const k = types.LayerID(25)           // the confidence interval // TODO: read from config
+
+type valueProvider interface {
+	Value(layer types.LayerID) (int, error)
+}
+
+type activeSetProvider interface {
+	GetActiveSetSize(layer types.LayerID) (uint32, error)
+}
+
+type Verifier interface {
+	Verify(msg, sig []byte) (bool, error)
+}
+
+type Oracle struct {
+	committeeSize uint32
+	beacon        valueProvider
+	asProvider    activeSetProvider
+	vrf           Verifier
+}
+
+func safeLayer(layer types.LayerID) types.LayerID {
+	if layer >= k {
+		return layer - k
+	}
+
+	return types.LayerID(genesisLayer)
```

Comment:
1. the case layer = k is already addressed (we return layer-k=0=genesis)
2. inside the `if' you return a number (layer - k), but outside it you return LayerID
3. genesisLayer := types.LayerID(0), so types.LayerID(genesisLayer) = types.LayerID(types.LayerID(0))

Reply:
Fixed